lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -440,6 +440,211 @@ class AutofillInfo {
   }
 }
 
+/// Replaces a range of text in the original string with the text given in the
+/// replacement string.
+String _replace(String originalText, String replacementText, ui.TextRange replacedRange) {
+  assert(replacedRange.isValid);
+  assert(replacedRange.start <= originalText.length && replacedRange.end <= originalText.length);
+
+  final ui.TextRange normalizedRange = ui.TextRange(start: math.min(replacedRange.start, replacedRange.end), end: math.max(replacedRange.start, replacedRange.end));
+
+  final String newText = normalizedRange.textBefore(originalText) + replacementText + normalizedRange.textAfter(originalText);
+
+  return newText;
+}
+
+/// The change between the last editing state and the current editing state
+/// of a text field.
+///
+/// This is packaged into a JSON and sent to the framework
+/// to be processed into a concrete [TextEditingDelta].
+class TextEditingDeltaState {
+  TextEditingDeltaState({
+    this.oldText = '',
+    this.deltaText = '',
+    this.deltaStart = -1,
+    this.deltaEnd = -1,
+    this.baseOffset = -1,
+    this.extentOffset = -1,
+    this.composingOffset = -1,
+    this.composingExtent = -1,
```

Comment:
Are the empty strings and -1 meaningful values, or are they here only to make the fields non-null? Would it make sense to mark the parameters with `required`?

Reply:
`required` wouldn't work in this case since we don't create the entire `TextEditingDelta` at any single point. It goes through `beforeInput` -> `onInput` -> `inferDeltaState`.

Reply:
On the framework side `-1` for `deltaStart` and `deltaEnd` is currently used by `TextEditingDelta` to identify a `NonTextUpdate`.  I made the `selection` and `composing` nullable now.